File: prototype/denoise/extend.py
"""
第2層：大域動きのみのカット間 拡張統合（docs/denoise_method_survey.md 2章）

パン・ズーム等の大域動きしかない区間では、隣接する保持グループの参照像を
大域ワープで現グループ座標に変換し、「一致する画素だけ」統合に参加させる。
これにより：

  - 1コマ撮り（毎フレーム動く）パン素材でも実効フレーム数Nを稼げる
    （第1層はグループ長1で無力なため、こうした素材の唯一の時間統合手段）
  - 2〜4コマ打ちの短いグループでも品質が第1層単独より上がる

安全設計（エッジを甘くしないための要点）：
  - 統合参加は画素単位の受け入れマスク制。ワープ後のトレランス付き残差が
    ノイズ床以下の画素だけが混ざる。セルが動いた領域・視差のある
    マルチプレーンレイヤー・ワープ誤差の大きいエッジは自動的に除外され、
    その画素は第1層の結果のまま残る（劣化しない）
  - 受け入れ率が低すぎる隣接グループ（＝実は絵が変わっている）は丸ごと棄却
  - ワープの信頼度が低い場合・移動量が大きすぎる場合も棄却
"""

from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np

from motion_classification.core import MotionThresholds, estimate_global_motion

from .core import DenoiseParams, _line_art_edge_mask, _luma


@dataclass
class ExtendParams:
    radius: int = 2                    # 前後それぞれ何グループまで統合に使うか
    accept_sigma_factor: float = 3.0   # 受け入れ閾値 = max(floor, factor×σ_pair)
    accept_floor: float = 3.0
    min_accept_ratio: float = 0.05     # 受け入れ画素率がこれ未満の隣接グループは棄却
    max_translation_ratio: float = 0.25  # 画面幅に対する並進がこれ超のワープは棄却
    min_warp_confidence: float = 0.2
    active_area_crop: float = 0.10     # 受け入れ率の算定・統計はフィルム枠を除いた中央で


def _estimate_warp_full(ref_src: np.ndarray, ref_dst: np.ndarray,
                        motion_thresholds: MotionThresholds) -> tuple[np.ndarray, float] | None:
    """ref_src → ref_dst 座標系への元解像度2x3ワープと信頼度を返す（失敗時 None）。"""
    src_u8 = np.clip(ref_src, 0, 255).astype(np.uint8)
    dst_u8 = np.clip(ref_dst, 0, 255).astype(np.uint8)
    g = estimate_global_motion(src_u8, dst_u8, motion_thresholds)
    if g.warp_work is None:
        return None
    warp = g.warp_work.copy().astype(np.float32)
    # workスケール→元解像度：線形部はそのまま、並進のみスケール
    to_orig = ref_src.shape[1] / motion_thresholds.work_width
    if to_orig > 1.0:
        warp[0, 2] *= to_orig
        warp[1, 2] *= to_orig
    return warp, g.confidence


def _tolerant_luma_residual(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """±1pxの位置ずれを許容した輝度残差（どちらも同座標系の画像）。"""
    ya = np.clip(_luma(a), 0, 255).astype(np.uint8)
    yb = np.clip(_luma(b), 0, 255).astype(np.uint8)
    k3 = np.ones((3, 3), np.uint8)
    over = cv2.subtract(yb, cv2.dilate(ya, k3)).astype(np.float32)
    under = cv2.subtract(cv2.erode(ya, k3), yb).astype(np.float32)
    return cv2.GaussianBlur(np.maximum(over, under), (5, 5), 0)


def extend_reference(center: dict, neighbors: list[dict],
                     params: ExtendParams | None = None,
                     denoise_params: DenoiseParams | None = None) -> dict:
    """中心グループの参照像を、隣接グループの参照像で拡張統合する。

    center / neighbors は analyze_hold_group の返り値
    （"reference"(float32) と "n" と "grain_sigma" を使う）。

    返り値：{
      "reference": 拡張後の参照像 (float32),
      "effective_n": 画素ごとの実効フレーム数マップ,
      "used_neighbors": 実際に統合に使えた隣接グループ数,
      "accept_ratios": 各隣接グループの受け入れ画素率,
    }
    """
    params = params or ExtendParams()
    denoise_params = denoise_params or DenoiseParams()

    ref_c = center["reference"].astype(np.float32)
    n_c = float(center["n"])
    sigma_c = max(center["grain_sigma"], 0.5)
    h, w = ref_c.shape[:2]

    motion_thresholds = MotionThresholds()

    acc = ref_c * n_c
    weight = np.full((h, w), n_c, np.float32)

    # エッジ画素は拡張統合から除外する。ワープには±1px程度の残留誤差があり、
    # トレランス受け入れをすり抜けた微小ずれエッジが混ざると鮮鋭度が落ちる
    # （実測：-4%）。グレイン除去の利益はほぼ平坦部にあるので、エッジは
    # 第1層（グループ内統合）の結果をそのまま保つのが最も安全
    not_edge = _line_art_edge_mask(np.clip(ref_c, 0, 255), dilate_px=5) == 0

    # 受け入れ率の算定はフィルム枠を除いた中央領域で行う
    my, mx = int(h * params.active_area_crop), int(w * params.active_area_crop)
    inner = np.zeros((h, w), bool)
    inner[my:h - my, mx:w - mx] = True

    used = 0
    accept_ratios = []

    for nb in neighbors:
        ref_n = nb["reference"].astype(np.float32)
        n_n = float(nb["n"])

        est = _estimate_warp_full(ref_n, ref_c, motion_thresholds)
        if est is None:
            accept_ratios.append(0.0)
            continue
        warp, confidence = est
        translation = float(np.hypot(warp[0, 2], warp[1, 2]))
        if confidence < params.min_warp_confidence or \
                translation > params.max_translation_ratio * w:
            accept_ratios.append(0.0)
            continue

        warped = cv2.warpAffine(ref_n, warp, (w, h),
                                flags=cv2.INTER_LANCZOS4,
                                borderMode=cv2.BORDER_CONSTANT,
                                borderValue=(-1000.0, -1000.0, -1000.0, 0.0))
        valid = (warped[..., 0] > -500.0)  # ワープで画面外から来た画素を除外

        resid = _tolerant_luma_residual(np.clip(warped, 0, 255), np.clip(ref_c, 0, 255))
        # 両参照像とも時間平均済みなのでノイズは σ/√N に減っている
        sigma_pair = sigma_c * float(np.sqrt(1.0 / max(n_c, 1) + 1.0 / max(n_n, 1)))
        threshold = max(params.accept_floor,
                        params.accept_sigma_factor * sigma_pair)

        accept = (resid <= threshold) & valid & not_edge
        # エッジ際のにじみ込みを防ぐため受け入れマスクを少し痩せさせる
        accept = cv2.erode(accept.astype(np.uint8), np.ones((3, 3), np.uint8)).astype(bool)

        ratio = float(accept[inner].mean())
        accept_ratios.append(round(ratio, 4))
        if ratio < params.min_accept_ratio:
            continue

        a = accept.astype(np.float32) * n_n
        acc += warped * a[..., None]
        weight += a
        used += 1

    reference_ext = acc / weight[..., None]

    return {
        "reference": reference_ext.astype(np.float32),
        "effective_n": weight,
        "used_neighbors": used,
        "accept_ratios": accept_ratios,
    }


def blend_spatial_fallback(image: np.ndarray, effective_n: np.ndarray,
                           grain_sigma: float, strength: float = 1.0) -> np.ndarray:
    """実効Nの低い画素にだけ空間NR（第3層）を混ぜる連続ブレンド。

    設計提案書3章「時間方向重みと空間方向重みを連続的に変化させるブレンド方式」の実装：
      - 実効N=1（時間統合が効かなかった画素）→ 空間NRを重み1.0で適用
      - 実効Nが増えるほど空間NRの重みを 1/N で減衰（時間統合で既にσ/√Nまで
        減っているため空間処理は不要になっていく）
    重みマップは空間的にぼかして分類境界のフリッカーを防ぐ（同章の注意点）。
    空間NR自体は線画エッジ保護つき（spatial_denoise_edge_preserving）なので、
    どの重みでもエッジには触れない。
    """
    if strength <= 0:
        return image
    from .core import spatial_denoise_edge_preserving

    u8 = np.clip(image, 0, 255).astype(np.uint8)
    denoised = spatial_denoise_edge_preserving(u8, grain_sigma, strength=strength)

    w = 1.0 / np.maximum(effective_n.astype(np.float32), 1.0)
    w = cv2.GaussianBlur(w, (31, 31), 0)[..., None]
    out = denoised.astype(np.float32) * w + u8.astype(np.float32) * (1.0 - w)
    return np.clip(out, 0, 255).astype(np.uint8)

File: analysis/src/extend.cpp
// 第2層：カット間拡張統合の C++ 実装。
// prototype/denoise/extend.py を仕様として忠実に移植する。

#include "animerestore/extend.h"

#include <cmath>

#include <opencv2/imgproc.hpp>

#include "animerestore/motion.h"

namespace animerestore {

namespace {

cv::Mat lumaU8(const cv::Mat& img32f3) {
    cv::Mat u8, gray;
    img32f3.convertTo(u8, CV_8UC3);
    cv::cvtColor(u8, gray, cv::COLOR_BGR2GRAY);
    return gray;
}

// ±1pxの位置ずれを許容した輝度残差（同座標系の2画像）
cv::Mat tolerantLumaResidual(const cv::Mat& a32f3, const cv::Mat& b32f3) {
    cv::Mat ya = lumaU8(a32f3), yb = lumaU8(b32f3);
    cv::Mat k3 = cv::Mat::ones(3, 3, CV_8U);
    cv::Mat dil, ero, over, under;
    cv::dilate(ya, dil, k3);
    cv::erode(ya, ero, k3);
    cv::subtract(yb, dil, over, cv::noArray(), CV_32F);
    cv::subtract(ero, yb, under, cv::noArray(), CV_32F);
    cv::Mat resid = cv::max(cv::max(over, under), 0.0f);
    cv::GaussianBlur(resid, resid, cv::Size(5, 5), 0);
    return resid;
}

}  // namespace

ExtendResult extendReference(const GroupAnalysis& center,
                             const std::vector<const GroupAnalysis*>& neighbors,
                             const ExtendParams& p) {
    const cv::Mat& refC = center.reference;
    const double nC = static_cast<double>(center.aligned.size());
    const double sigmaC = std::max(center.grainSigma, 0.5);
    const int h = refC.rows, w = refC.cols;

    ExtendResult r;
    cv::Mat acc = refC * nC;
    r.effectiveN = cv::Mat(h, w, CV_32F, cv::Scalar(nC));

    // エッジ除外マスクは denoise 側の XDoG 実装を共有（Python extend.py と同値。
    // 従来ここだけ Canny のままでパリティが逸脱していた＝監査で発見・修正）
    cv::Mat notEdge = ~lineArtEdgeMask(refC, 5);

    int my = static_cast<int>(h * p.activeAreaCrop);
    int mx = static_cast<int>(w * p.activeAreaCrop);
    cv::Rect inner(mx, my, w - 2 * mx, h - 2 * my);

    MotionThresholds mth;
    cv::Mat refCu8;
    refC.convertTo(refCu8, CV_8UC3);

    for (const GroupAnalysis* nb : neighbors) {
        const cv::Mat& refN = nb->reference;
        const double nN = static_cast<double>(nb->aligned.size());

        cv::Mat refNu8;
        refN.convertTo(refNu8, CV_8UC3);
        GlobalMotion g = estimateGlobalMotion(refNu8, refCu8, mth);
        if (g.warpWork.empty()) {
            r.acceptRatios.push_back(0.0);
            continue;
        }
        // workスケール→元解像度：並進のみスケール
        cv::Mat warp = g.warpWork.clone();
        double toOrig = static_cast<double>(w) / mth.workWidth;
        if (toOrig > 1.0) {
            warp.at<float>(0, 2) *= static_cast<float>(toOrig);
            warp.at<float>(1, 2) *= static_cast<float>(toOrig);
        }
        double translation = std::hypot(warp.at<float>(0, 2), warp.at<float>(1, 2));
        if (g.confidence < p.minWarpConfidence ||
            translation > p.maxTranslationRatio * w) {
            r.acceptRatios.push_back(0.0);
            continue;
        }

        cv::Mat warped;
        cv::warpAffine(refN, warped, warp, cv::Size(w, h), cv::INTER_LANCZOS4,
                       cv::BORDER_CONSTANT, cv::Scalar(-1000, -1000, -1000));
        cv::Mat channels[3];
        cv::split(warped, channels);
        cv::Mat valid = channels[0] > -500.0f;

        cv::Mat warpedClip;
        cv::max(warped, 0.0f, warpedClip);
        cv::min(warpedClip, 255.0f, warpedClip);
        cv::Mat resid = tolerantLumaResidual(warpedClip, refC);
        double sigmaPair = sigmaC * std::sqrt(1.0 / std::max(nC, 1.0) +
                                              1.0 / std::max(nN, 1.0));
        double threshold = std::max(p.acceptFloor, p.acceptSigmaFactor * sigmaPair);

        cv::Mat accept = (resid <= threshold) & valid & notEdge;
        cv::erode(accept, accept, cv::Mat::ones(3, 3, CV_8U));

        double ratio = static_cast<double>(cv::countNonZero(accept(inner))) /
                       inner.area();
        r.acceptRatios.push_back(ratio);
        if (ratio < p.minAcceptRatio) continue;

        cv::Mat aW;
        accept.convertTo(aW, CV_32F, nN / 255.0);
        cv::Mat aW3;
        cv::cvtColor(aW, aW3, cv::COLOR_GRAY2BGR);
        acc += warpedClip.mul(aW3);
        r.effectiveN += aW;
        r.usedNeighbors++;
    }

    cv::Mat n3;
    cv::cvtColor(r.effectiveN, n3, cv::COLOR_GRAY2BGR);
    r.reference = acc / n3;
    return r;
}

cv::Mat blendSpatialFallback(const cv::Mat& image, const cv::Mat& effectiveN,
                             double grainSigma, double strength) {
    if (strength <= 0) return image;
    cv::Mat u8;
    if (image.type() == CV_8UC3) u8 = image;
    else image.convertTo(u8, CV_8UC3);

    cv::Mat denoised = spatialDenoiseEdgePreserving(u8, grainSigma, strength);

    cv::Mat w;
    cv::max(effectiveN, 1.0f, w);
    cv::divide(1.0, w, w);
    cv::GaussianBlur(w, w, cv::Size(31, 31), 0);
    cv::Mat w3;
    cv::cvtColor(w, w3, cv::COLOR_GRAY2BGR);

    cv::Mat fD, fO;
    denoised.convertTo(fD, CV_32FC3);
    u8.convertTo(fO, CV_32FC3);
    cv::Mat out = fD.mul(w3) + fO.mul(cv::Scalar::all(1.0) - w3);
    out.convertTo(out, CV_8UC3);
    return out;
}

}  // namespace animerestore
